basic/geometry_math.go

```diff
@@ -6,151 +6,202 @@ import (
-	case tegola.Point:
+		return nil, fmt.Errorf("Unknown Geometry: %+v", geometry)
+
+	case geom.Point:
 		c, err := f(geo.X(), geo.Y())
 		if err != nil {
-			return G{}, err
+			return nil, err
 		}
 		if len(c) < 2 {
-			return G{}, fmt.Errorf("Function did not return minimum number of coordinates got %v expected 2", len(c))
-		}
-		return G{Point{c[0], c[1]}}, nil
-	case tegola.Point3:
-		c, err := f(geo.X(), geo.Y(), geo.Z())
-		if err != nil {
-			return G{}, err
-		}
-		if len(c) < 3 {
-			return G{}, fmt.Errorf("Function did not return minimum number of coordinates got %v expected 3", len(c))
+			return nil, fmt.Errorf("function did not return minimum number of coordinates got %v expected 2", len(c))
 		}
-		return G{Point3{c[0], c[1], c[2]}}, nil
-	case tegola.MultiPoint:
-		var pts MultiPoint
-		for _, pt := range geo.Points() {
-			c, err := f(pt.X(), pt.Y())
+		return geom.Point{c[0], c[1]}, nil
+
+	case geom.MultiPoint:
+		pts := make(geom.MultiPoint, len(geo))
+
+		for i, pt := range geo {
+			c, err := f(pt[:]...)
 			if err != nil {
-				return G{}, err
+				return nil, err
 			}
 			if len(c) < 2 {
-				return G{}, fmt.Errorf("Function did not return minimum number of coordinates got %v expected 2", len(c))
+				return nil, fmt.Errorf("function did not return minimum number of coordinates got %v expected 2", len(c))
 			}
-			pts = append(pts, Point{c[0], c[1]})
+			pts[i][0], pts[i][1] = c[0], c[1]
 		}
-		return G{pts}, nil
-	case tegola.LineString:
-		var line Line
-		for _, ptGeo := range geo.Subpoints() {
-			c, err := f(ptGeo.X(), ptGeo.Y())
+		return pts, nil
+
+	case geom.LineString:
+		line := make(geom.LineString, len(geo))
+		for i, pt := range geo {
+			c, err := f(pt[:]...)
 			if err != nil {
-				return G{}, err
+				return nil, err
 			}
 			if len(c) < 2 {
-				return G{}, fmt.Errorf("Function did not return minimum number of coordinates got %v expected 2", len(c))
+				return nil, fmt.Errorf("function did not return minimum number of coordinates got %v expected 2", len(c))
 			}
-			line = append(line, Point{c[0], c[1]})
+			line[i][0], line[i][1] = c[0], c[1]
 		}
-		return G{line}, nil
-	case tegola.MultiLine:
-		var line MultiLine
-		for i, lineGeo := range geo.Lines() {
-			geoLine, err := ApplyToPoints(lineGeo, f)
+		return line, nil
+
+	case geom.MultiLineString:
+		lines := make(geom.MultiLineString, len(geo))
+
+		for i, line := range geo {
+			// getting a geometry inteface back
+			linei, err := ApplyToPoints(geom.LineString(line), f)
 			if err != nil {
-				return G{}, fmt.Errorf("Got error converting line(%v) of multiline: %v", i, err)
+				return nil, fmt.Errorf("got error converting line(%v) of multiline: %v", i, err)
 			}
-			if !geoLine.IsLine() {
-				panic("We did not get the conversion we were expecting")
+
+			// get the value
+			linev, ok := linei.(geom.LineString)
+			if !ok {
+				panic("we did not get the conversion we were expecting")
```

Comment:
@gdey do you think we should keep the panic here or return an error?

Reply:
If it gets to this point something with the code is wrong. And nothing we can do is recoverable. So, I think it should panic.